### PR TITLE
MAINT Fix lint, clean and expand docstrings

### PR DESCRIPTION
**Pull request by lucyleeow**
Addresses some docstring lint rules previously ignored.

* Ignore "D103" (Missing docstring in public function) for tests - there were many tests without docstrings and it did not seem necessary to add these.
* This PR should only include docstring changes - other changes will be in other PRs

Turned out to be larger than I initially expected. If you're wanting to take a look @larsoner, would be useful to just look at any 'parameter' additions to make sure they are correct.

### CODE REVIEW

**Review by larsoner — APPROVED** [review]
Awesome @lucyleeow , I was not happy adding all those `ignores` -- it'll be nice to have them gone!

File: sphinx_gallery/scrapers.py [hunk 1]
```diff
@@ -455,13 +453,14 @@ def figure_rst(figure_list, sources_dir, fig_titles="", srcsetpaths=None):
...
-    hinames posix paths to
+    """Create the srcset string for including on the rst line.
+
+    For example; `sources_dir` might be `/home/sample-proj/source`,
+    hinames posix paths:
     0: /home/sample-proj/source/plot_types/images/img1.png,
     2.0: /home/sample-proj/source/plot_types/images/img1_2_00x.png,
-    The result will be:
+
+    The result wwould be:
```
Comment (larsoner) [inline sphinx_gallery/scrapers.py:463]:
```suggestion
    The result would be:
```
Also there are spelling failures in CIs

File: sphinx_gallery/tests/tinybuild/examples/plot_repr.py [hunk 2]
```diff
@@ -7,8 +7,14 @@
...
+
     def _repr_html_(self):
         return "<p><b>This should print<b></p>"
 
 
 A()
+
+import numpy as np
+
+np.arange(6)
```
Comment (larsoner) [inline sphinx_gallery/tests/tinybuild/examples/plot_repr.py:20]: Why add this in this PR, perhaps better as a separate one?

Comment (larsoner) [reply sphinx_gallery/tests/tinybuild/examples/plot_repr.py:20]: ... might be the cause of this failure for example

https://dev.azure.com/sphinx-gallery/sphinx-gallery/_build/results?buildId=1451&view=logs&j=66ad8d10-cb69-57fe-8782-e0500ab4b7b6&t=b2a44cde-8fb7-5f37-5248-a7b5141501c3&l=264

Comment (lucyleeow) [reply sphinx_gallery/tests/tinybuild/examples/plot_repr.py:20]: sorry leftover cruft

File: sphinx_gallery/backreferences.py [hunk 3]
```diff
@@ -222,7 +238,34 @@ def _make_ref_regex(config=None):
...
+    ----------
+    script_blocks : list
+        (label, content, line_number)
+        List where each element is a tuple with the label ('text' or 'code'),
+        the corresponding content string of block and the leading line number.
+    ref_regex : str
+        Regex to find references to python objects.
+    example_globals: Optional[Dict[str, Any]]
+        Global variables for examples. Default=None
+    node : ast.Module or str
```
Comment (lucyleeow) [inline sphinx_gallery/backreferences.py:253]: When using default 'exec' I think the output is `ast.Module`:
https://greentreesnakes.readthedocs.io/en/latest/tofrom.html#getting-to-and-from-asts

**Comment by lucyleeow** [comment]
Thanks! I did not expect CIs to be green already!

**Pull request merged**